### PR TITLE
add imenu support

### PR DESCRIPTION
**Pull request by Junker**

### CODE REVIEW

File: vala-mode.el [hunk 1]
```diff
@@ -352,6 +354,10 @@
   :type 'hook
   :group 'c)
 
+(defcustom vala-want-imenu t
```
Comment (purcell) [inline vala-mode.el]: This is not really idiomatic, and I'd suggest skipping it. Typical major modes never auto-enable imenu. Instead, they rely on the user adding something like `imenu-add-menubar-index` to the mode hook (or `prog-mode-hook` if they really love `imenu`).

File: vala-mode.el [hunk 2]
```diff
@@ -393,7 +399,381 @@ Key bindings:
   (c-toggle-hungry-state -1)
   (run-hooks 'c-mode-common-hook)
   (run-hooks 'vala-mode-hook)
-  (c-update-modeline))
+  (c-update-modeline)
+
+  ;; maybe do imenu scan after hook returns
+  (when vala-want-imenu
+    (vala--setup-imenu)))
```
Comment (purcell) [inline vala-mode.el]: So yeah, remove this bit, but set `imenu-create-index-function` here unconditionally.

File: vala-mode.el [hunk 3]
```diff
@@ -352,6 +354,10 @@
   :type 'hook
   :group 'c)
 
+(defcustom vala-want-imenu t
+  "*Whether to generate a buffer index via imenu for Vala buffers."
+  :type 'boolean :group 'vala)
+
 ;;; The entry point into the mode
 ;;;###autoload
 (defun vala-mode ()
```
Comment (purcell) [inline vala-mode.el]: As an aside, converting this to `define-derived-mode` would be a good plan here, and allow a bunch of code to be removed. `prog-mode` would be the appropriate parent mode.

File: vala-mode.el [hunk 4]
```diff
@@ -393,7 +399,381 @@ Key bindings:
...
+    ;; (Enums always sort last though, because they don't have
+    ;; sub-menus)
+    (vala--imenu-sort result)))
+
+(defun vala--imenu-create-index-function ()
+  "Create an imenu index."
+  (vala--imenu-transform-index
+   (imenu--generic-function vala--imenu-expression)))
+
+(defun vala--setup-imenu ()
```
Comment (purcell) [inline vala-mode.el]: You would then also remove this function. :-)

**Comment by tarsius** [comment]
> I left a couple of small suggestions.

Thanks!

**Comment by tarsius** [comment]
Closing due to lack of activity.

**Pull request closed without merging**